### PR TITLE
Add types to Component classes

### PR DESCRIPTION
**Pull request by philippjfr**
- [x] lumen/base.py
- [x] lumen/dashboard.py
- [x] lumen/views/base.py

### CODE REVIEW

File: lumen/views/base.py [hunk 1]
```diff
@@ -269,7 +268,7 @@ def _validate_pipeline(cls, *args, **kwargs):
     ##################################################################
 
     @classmethod
-    def from_spec(cls, spec: Dict[str, Any], pipeline=None, source=None, filters=None):
+    def from_spec(cls, spec: Dict[str, Any], source=None, filters=None, pipeline=None):
```
Comment (Hoxbro) [inline lumen/views/base.py:271]: Should we make it keywords-only argument after spec?

Comment (philippjfr) [reply lumen/views/base.py]: Yes, seems like a good idea.

Comment (philippjfr) [reply lumen/views/base.py]: Can you file an issue?

File: lumen/auth.py [hunk 2]
```diff
@@ -12,10 +17,10 @@ class AuthPlugin(param.Parameterized):
     transforms to it.
     """
 
-    auth_type = None
+    auth_type: ClassVar[str | None] = None
 
     @classmethod
-    def _get_type(cls, auth_type):
+    def _get_type(cls, auth_type: str) -> Type['AuthPlugin']:
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply lumen/auth.py:23]: Self is not available in 3.7 I think.

[the remark this answers is not in the document]
Comment (maximlt) [reply lumen/auth.py:23]: Hmm yeah I actually don't know where to find which typing features has been backported and until where. Thought people were using https://github.com/python/typing_extensions to get those features, and that is available from Python 3.7.

[the remark this answers is not in the document]
Comment (philippjfr) [reply lumen/auth.py:23]: Was not able to get this working:

```
lumen/auth.py:25: error: Variable "typing_extensions.Self" is not valid as a type
lumen/auth.py:25: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
lumen/auth.py:41: error: Cannot instantiate type "Type[Self?]"
```

**Pull request merged**